File: lotrs-py/ring.py
"""
ring.py -- Polynomial ring  R_q = Z_q[X] / (X^d + 1).

Coefficients are Python ints in [0, q) (unsigned canonical form).
Signed / centred form [-q//2, q//2] is used where noted.

Two multiplication paths:
  * Schoolbook negacyclic convolution  (always available; correctness
    reference for tests).
  * CRT-NTT backend via two 48-bit auxiliary primes  (see aux_ntt.py).
    Enabled when d in {128, 256}.  The scheme modulus q need not be
    NTT-friendly; fast multiplication is carried by the auxiliary primes
    and reconstruction is exact because P = p1 * p2 ≈ 2^96 exceeds every
    possible integer product coefficient for the sizes of q, q_hat, d
    used in the scheme.

Vectors of ring elements are plain Python lists; matrices are lists of
rows (each row itself a list of ring elements).
"""

from aux_ntt import CRTBackend, SUPPORTED_D


class Ring:
    """Arithmetic in  R_q = Z_q[X] / (X^d + 1)."""

    def __init__(self, q: int, d: int):
        self.q = q
        self.d = d
        self.half_q = q // 2
        self._ntt_ctx = None
        if d in SUPPORTED_D:
            self._ntt_ctx = CRTBackend(q, d)

    # ---- element creation ------------------------------------------------

    def zero(self):
        """Additive identity."""
        return [0] * self.d

    def one(self):
        """Multiplicative identity (constant 1)."""
        r = [0] * self.d
        r[0] = 1
        return r

    def const(self, c):
        """Constant polynomial  c mod q."""
        r = [0] * self.d
        r[0] = int(c) % self.q
        return r

    # ---- coefficient form conversions ------------------------------------

    def reduce(self, a):
        """Reduce every coefficient into [0, q)."""
        q = self.q
        return [int(c) % q for c in a]

    def centered(self, a):
        """Unsigned [0,q) --> signed [-q//2, q//2]."""
        h, q = self.half_q, self.q
        return [c - q if c > h else c for c in a]

    def from_centered(self, a):
        """Signed --> unsigned [0, q)."""
        q = self.q
        return [c % q for c in a]

    # ---- element-wise arithmetic -----------------------------------------

    def add(self, a, b):
        q = self.q
        return [(a[i] + b[i]) % q for i in range(self.d)]

    def sub(self, a, b):
        q = self.q
        return [(a[i] - b[i]) % q for i in range(self.d)]

    def neg(self, a):
        q = self.q
        return [(q - c) % q for c in a]

    def scale(self, c, a):
        """Integer scalar  c  times polynomial  a."""
        q = self.q
        c = int(c) % q
        return [(c * ai) % q for ai in a]

    def mul(self, a, b):
        """Negacyclic convolution  a * b  in R_q."""
        if self._ntt_ctx is not None:
            return self._ntt_ctx.mul(a, b)
        return self._mul_schoolbook(a, b)

    def _mul_schoolbook(self, a, b):
        """Schoolbook negacyclic convolution (reference implementation)."""
        d, q = self.d, self.q
        c = [0] * d
        for i in range(d):
            ai = a[i]
            if ai == 0:
                continue
            for j in range(d):
                k = i + j
                if k < d:
                    c[k] = (c[k] + ai * b[j]) % q
                else:
                    c[k - d] = (c[k - d] - ai * b[j]) % q
        return c

    # ---- norms (operate on unsigned-canonical coefficients) ---------------

    def inf_norm(self, a):
        """||a||_inf of the centred representation."""
        h, q = self.half_q, self.q
        return max((q - c if c > h else c) for c in a)

    def l2_norm_sq(self, a):
        """||a||_2^2 of the centred representation."""
        h, q = self.half_q, self.q
        return sum((c - q) ** 2 if c > h else c * c for c in a)

    def l1_norm(self, a):
        """||a||_1 of the centred representation."""
        h, q = self.half_q, self.q
        return sum((q - c if c > h else c) for c in a)

    # ---- vector operations -----------------------------------------------
    # A "vector" is  list[poly]  where  poly = list[int].

    def vec_zero(self, n):
        return [self.zero() for _ in range(n)]

    def vec_add(self, u, v):
        return [self.add(u[i], v[i]) for i in range(len(u))]

    def vec_sub(self, u, v):
        return [self.sub(u[i], v[i]) for i in range(len(u))]

    def vec_neg(self, v):
        return [self.neg(vi) for vi in v]

    def vec_scale(self, c_poly, v):
        """Ring-element scalar  c_poly  times vector  v."""
        return [self.mul(c_poly, vi) for vi in v]

    def vec_scale_int(self, c, v):
        """Integer scalar  c  times vector  v."""
        return [self.scale(c, vi) for vi in v]

    def inner(self, u, v):
        """<u, v> = sum_i u_i * v_i  in R_q."""
        acc = self.zero()
        for i in range(len(u)):
            acc = self.add(acc, self.mul(u[i], v[i]))
        return acc

    def vec_inf_norm(self, v):
        if not v:
            return 0
        return max(self.inf_norm(vi) for vi in v)

    def vec_l2_norm_sq(self, v):
        return sum(self.l2_norm_sq(vi) for vi in v)

    # ---- matrix operations -----------------------------------------------
    # A "matrix" is  list[list[poly]]  (row-major).

    def mat_vec(self, M, v):
        """Matrix-vector product  M * v."""
        return [self.inner(row, v) for row in M]

    # ---- helpers ---------------------------------------------------------

    def vec_concat(self, *vecs):
        out = []
        for v in vecs:
            out.extend(v)
        return out

    def centered_decompose(self, a, K):
        """
        Coefficient-wise centred decomposition.

        For each coefficient c (in centred form), write
            c = high * 2^K  +  low     with  |low| <= 2^{K-1}.

        Returns (high_poly, low_poly) both in unsigned canonical form.
        """
        q, d = self.q, self.d
        h = self.half_q
        power = 1 << K
        half_power = power >> 1
        hi = [0] * d
        lo = [0] * d
        for i in range(d):
            c = a[i] - q if a[i] > h else a[i]      # signed
            # Python's divmod rounds toward -inf, so use explicit formula
            low = c % power
            if low > half_power:
                low -= power
            high = (c - low) // power
            hi[i] = high % q
            lo[i] = low % q
        return hi, lo


# --------------------------------------------------------------------------
if __name__ == "__main__":
    R = Ring(7681, 32)

    # basic identities
    a = R.one()
    z = R.zero()
    assert R.mul(a, a) == a, "1*1 == 1"
    assert R.add(z, a) == a, "0+1 == 1"

    # negacyclic property:  X^d == -1
    x_d = [0] * 32
    x_d[0] = 7680            # -1 mod 7681
    x_1 = [0] * 32
    x_1[1] = 1               # X
    prod = x_1
    for _ in range(31):       # X^32
        prod = R.mul(prod, x_1)
    assert prod == x_d, "X^d == -1 in R_q"

    # commutativity / associativity spot-check
    import random
    rng = random.Random(42)
    def rand_poly():
        return [rng.randrange(7681) for _ in range(32)]

    a, b, c = rand_poly(), rand_poly(), rand_poly()
    assert R.mul(a, b) == R.mul(b, a), "commutativity"
    assert R.mul(R.mul(a, b), c) == R.mul(a, R.mul(b, c)), "associativity"
    assert R.mul(a, R.add(b, c)) == R.add(R.mul(a, b), R.mul(a, c)), \
        "distributivity"

    # centred decomposition round-trip
    p = rand_poly()
    hi, lo = R.centered_decompose(p, 5)
    recon = R.add(R.scale(32, hi), lo)
    assert recon == p, "centred decomposition round-trip"

    print("ring.py: all self-tests passed")

File: lotrs-rs/src/ring.rs
//! Polynomial ring `R_q = Z_q[X]/(X^d + 1)` — port of `lotrs-py/ring.py`.
//!
//! Coefficients are unsigned `u64` in `[0, q)` (canonical form).  Signed
//! / centred form `[-q/2, q/2]` is used where noted (returned as `i64`).
//!
//! Two multiplication paths:
//!
//! * **Schoolbook** — O(d²) reference, always available.
//! * **CRT-NTT** — enabled when `d ∈ {128, 256}`, see [`crate::aux_ntt`].
//!
//! Polynomials are plain `Vec<u64>` (or `&[u64]` / `&mut [u64]` in
//! signatures).  Vectors of polynomials are `Vec<Vec<u64>>`; matrices
//! are row-major `Vec<Vec<Vec<u64>>>`.

use crate::aux_ntt::{CrtBackend, CrtNttMat};

pub type Poly = Vec<u64>;
pub type PolyVec = Vec<Poly>;
pub type PolyMat = Vec<PolyVec>;
pub type NttPolyMat = CrtNttMat;

/// Arithmetic context for `R_q`.  Cheap to construct when the NTT fast
/// path does not apply; O(d) memory + O(d) precompute otherwise.
#[derive(Clone)]
pub struct Ring {
    pub q: u64,
    pub d: usize,
    pub half_q: u64,
    backend: Option<CrtBackend>,
}

impl Ring {
    pub fn new(q: u64, d: usize) -> Self {
        Self {
            q,
            d,
            half_q: q / 2,
            backend: CrtBackend::new(q, d),
        }
    }

    // ---- element creation -----------------------------------------------

    #[inline]
    pub fn zero(&self) -> Poly {
        vec![0u64; self.d]
    }

    #[inline]
    pub fn one(&self) -> Poly {
        let mut r = vec![0u64; self.d];
        r[0] = 1;
        r
    }

    pub fn const_poly(&self, c: i64) -> Poly {
        let mut r = vec![0u64; self.d];
        r[0] = (c.rem_euclid(self.q as i64)) as u64;
        r
    }

    // ---- coefficient form conversions -----------------------------------

    pub fn reduce(&self, a: &[u64]) -> Poly {
        a.iter().map(|&c| c % self.q).collect()
    }

    /// Unsigned `[0, q)` → signed `[-q/2, q/2]`.
    pub fn centered(&self, a: &[u64]) -> Vec<i64> {
        let q = self.q as i64;
        let h = self.half_q;
        a.iter()
            .map(|&c| if c > h { c as i64 - q } else { c as i64 })
            .collect()
    }

    /// Signed → unsigned `[0, q)`.
    pub fn from_centered(&self, a: &[i64]) -> Poly {
        let q = self.q as i64;
        a.iter().map(|&c| c.rem_euclid(q) as u64).collect()
    }

    // ---- element-wise arithmetic ---------------------------------------

    #[inline]
    fn add_coeff(&self, a: u64, b: u64) -> u64 {
        let q = self.q;
        let s = a + b;
        let mask = 0u64.wrapping_sub((s >= q) as u64);
        s.wrapping_sub(q & mask)
    }

    #[inline]
    fn sub_coeff(&self, a: u64, b: u64) -> u64 {
        let q = self.q;
        let d = a.wrapping_sub(b);
        let mask = 0u64.wrapping_sub((a < b) as u64);
        d.wrapping_add(q & mask)
    }

    #[inline]
    fn neg_coeff(&self, c: u64) -> u64 {
        let q = self.q;
        let zmask = 0u64.wrapping_sub((c != 0) as u64);
        q.wrapping_sub(c) & zmask
    }

    pub fn add(&self, a: &[u64], b: &[u64]) -> Poly {
        (0..self.d).map(|i| self.add_coeff(a[i], b[i])).collect()
    }

    pub fn sub(&self, a: &[u64], b: &[u64]) -> Poly {
        (0..self.d).map(|i| self.sub_coeff(a[i], b[i])).collect()
    }

    pub fn neg(&self, a: &[u64]) -> Poly {
        a.iter().map(|&c| self.neg_coeff(c)).collect()
    }

    pub fn scale(&self, c: i64, a: &[u64]) -> Poly {
        let q = self.q;
        let c_mod = c.rem_euclid(q as i64) as u64;
        a.iter()
            .map(|&ai| ((c_mod as u128 * ai as u128) % q as u128) as u64)
            .collect()
    }

    pub fn mul(&self, a: &[u64], b: &[u64]) -> Poly {
        if let Some(ref bk) = self.backend {
            bk.mul(a, b)
        } else {
            self.mul_schoolbook(a, b)
        }
    }

    /// Schoolbook negacyclic convolution — always available, used by
    /// tests as the correctness oracle.
    pub fn mul_schoolbook(&self, a: &[u64], b: &[u64]) -> Poly {
        let d = self.d;
        let q = self.q as i128;
        let mut c = vec![0i128; d];
        for i in 0..d {
            let ai = a[i] as i128;
            if ai == 0 {
                continue;
            }
            for j in 0..d {
                let prod = ai * b[j] as i128;
                let k = i + j;
                if k < d {
                    c[k] = (c[k] + prod) % q;
                } else {
                    c[k - d] = (c[k - d] - prod) % q;
                }
            }
        }
        c.into_iter().map(|v| v.rem_euclid(q) as u64).collect()
    }

    // ---- norms (on centred representation) ------------------------------

    pub fn inf_norm(&self, a: &[u64]) -> u64 {
        let h = self.half_q;
        let q = self.q;
        a.iter()
            .map(|&c| if c > h { q - c } else { c })
            .max()
            .unwrap_or(0)
    }

    pub fn l2_norm_sq(&self, a: &[u64]) -> u128 {
        let h = self.half_q;
        let q = self.q;
        a.iter()
            .map(|&c| {
                let cc = if c > h { (q - c) as u128 } else { c as u128 };
                cc * cc
            })
            .sum()
    }

    pub fn l1_norm(&self, a: &[u64]) -> u128 {
        let h = self.half_q;
        let q = self.q;
        a.iter()
            .map(|&c| if c > h { (q - c) as u128 } else { c as u128 })
            .sum()
    }

    // ---- vector operations ---------------------------------------------

    pub fn vec_zero(&self, n: usize) -> PolyVec {
        (0..n).map(|_| self.zero()).collect()
    }

    pub fn vec_add(&self, u: &[Poly], v: &[Poly]) -> PolyVec {
        u.iter()
            .zip(v.iter())
            .map(|(a, b)| self.add(a, b))
            .collect()
    }

    pub fn vec_sub(&self, u: &[Poly], v: &[Poly]) -> PolyVec {
        u.iter()
            .zip(v.iter())
            .map(|(a, b)| self.sub(a, b))
            .collect()
    }

    pub fn vec_neg(&self, v: &[Poly]) -> PolyVec {
        v.iter().map(|p| self.neg(p)).collect()
    }

    pub fn vec_scale(&self, c_poly: &[u64], v: &[Poly]) -> PolyVec {
        v.iter().map(|p| self.mul(c_poly, p)).collect()
    }

    pub fn vec_scale_int(&self, c: i64, v: &[Poly]) -> PolyVec {
        v.iter().map(|p| self.scale(c, p)).collect()
    }

    // ---- in-place / fused helpers --------------------------------------
    // These exist to cut allocation churn on the sign / verify hot paths.
    // The generic public vec_* functions stay (used widely), but the
    // BENCH-dominating loops in sign1 / kagg switch to in-place variants.

    /// `a += b` in place.
    #[inline]
    pub fn add_assign(&self, a: &mut [u64], b: &[u64]) {
        for i in 0..self.d {
            a[i] = self.add_coeff(a[i], b[i]);
        }
    }

    /// `a -= b` in place.
    #[inline]
    pub fn sub_assign(&self, a: &mut [u64], b: &[u64]) {
        for i in 0..self.d {
            a[i] = self.sub_coeff(a[i], b[i]);
        }
    }

    /// `u += v` element-wise in place.
    pub fn vec_add_assign(&self, u: &mut [Poly], v: &[Poly]) {
        for (a, b) in u.iter_mut().zip(v.iter()) {
            self.add_assign(a, b);
        }
    }

    /// `u -= v` element-wise in place.
    pub fn vec_sub_assign(&self, u: &mut [Poly], v: &[Poly]) {
        for (a, b) in u.iter_mut().zip(v.iter()) {
            self.sub_assign(a, b);
        }
    }

    /// `acc += c · v` element-wise (ring multiply then add).  Fused helper
    /// used by `kagg` and the `pk_term` loop inside `sign1`.
    pub fn vec_add_scaled(&self, acc: &mut [Poly], c: &[u64], v: &[Poly]) {
        for i in 0..v.len() {
            let prod = self.mul(c, &v[i]);
            self.add_assign(&mut acc[i], &prod);
        }
    }

    /// `acc -= c · v` element-wise.
    pub fn vec_sub_scaled(&self, acc: &mut [Poly], c: &[u64], v: &[Poly]) {
        for i in 0..v.len() {
            let prod = self.mul(c, &v[i]);
            self.sub_assign(&mut acc[i], &prod);
        }
    }

    pub fn inner(&self, u: &[Poly], v: &[Poly]) -> Poly {
        let mut acc = self.zero();
        for i in 0..u.len() {
            let prod = self.mul(&u[i], &v[i]);
            self.add_assign(&mut acc, &prod);
        }
        acc
    }

    pub fn vec_inf_norm(&self, v: &[Poly]) -> u64 {
        v.iter().map(|p| self.inf_norm(p)).max().unwrap_or(0)
    }

    pub fn vec_l2_norm_sq(&self, v: &[Poly]) -> u128 {
        v.iter().map(|p| self.l2_norm_sq(p)).sum()
    }

    pub fn mat_vec(&self, m: &[PolyVec], v: &[Poly]) -> PolyVec {
        m.iter().map(|row| self.inner(row, v)).collect()
    }

    /// Pre-transform a matrix for repeated CRT-NTT matrix-vector
    /// products.  Returns `None` when the ring is using schoolbook
    /// multiplication.
    pub fn mat_to_ntt(&self, m: &[PolyVec]) -> Option<NttPolyMat> {
        self.backend.as_ref().map(|bk| bk.mat_to_ntt(m))
    }

    /// Matrix-vector product using a pre-transformed matrix.  Returns
    /// `None` when the ring has no CRT backend.
    pub fn mat_vec_ntt(&self, m: &NttPolyMat, v: &[Poly]) -> Option<PolyVec> {
        self.backend.as_ref().map(|bk| bk.mat_vec_ntt(m, v))
    }

    pub fn vec_concat(&self, parts: &[&[Poly]]) -> PolyVec {
        let mut out = Vec::new();
        for p in parts {
            out.extend_from_slice(p);
        }
        out
    }

    // ---- centred decomposition (Bai-Galbraith) --------------------------

    /// Split each centred coefficient `c` as `hi * 2^K + lo` with
    /// `lo ∈ (-2^{K-1}, 2^{K-1}]`.  Returns `(hi, lo)` both in unsigned
    /// canonical form.
    ///
    /// The split matches `lotrs-py/ring.py::Ring.centered_decompose`
    /// exactly — in particular, when `c ≡ 2^{K-1} (mod 2^K)` the low
    /// part is `+2^{K-1}` (not `-2^{K-1}`), which affects the value of
    /// `hi` and therefore the Fiat-Shamir hash.
    pub fn centered_decompose(&self, a: &[u64], k_bits: u32) -> (Poly, Poly) {
        let q = self.q as i64;
        let h = self.half_q as i64;
        let step = 1i64 << k_bits;
        let half_step = 1i64 << (k_bits - 1);

        let mut hi = vec![0u64; self.d];
        let mut lo = vec![0u64; self.d];

        for i in 0..self.d {
            let c = a[i] as i64;
            let c_centred = if c > h { c - q } else { c };
            // Python:  low = c % power; if low > half_power: low -= power
            let mut low = c_centred.rem_euclid(step);
            if low > half_step {
                low -= step;
            }
            let high = (c_centred - low) / step;
            lo[i] = low.rem_euclid(q) as u64;
            hi[i] = high.rem_euclid(q) as u64;
        }
        (hi, lo)
    }
}

// -------------------------------------------------------------------------

#[cfg(test)]
mod tests {
    use super::*;
    use crate::params::{PRODUCTION_PARAMS, TEST_PARAMS};

    fn lcg(mut s: u64) -> impl FnMut(u64) -> u64 {
        move |bound: u64| {
            s = s
                .wrapping_mul(6_364_136_223_846_793_005)
                .wrapping_add(1_442_695_040_888_963_407);
            (s >> 16) % bound
        }
    }

    #[test]
    fn commutativity_d32() {
        let R = Ring::new(TEST_PARAMS.q, 32);
        let mut r = lcg(1);
        let a: Poly = (0..32).map(|_| r(R.q)).collect();
        let b: Poly = (0..32).map(|_| r(R.q)).collect();
        assert_eq!(R.mul(&a, &b), R.mul(&b, &a));
    }

    #[test]
    fn multiplicative_identity() {
        let R = Ring::new(TEST_PARAMS.q, 32);
        let mut r = lcg(2);
        let a: Poly = (0..32).map(|_| r(R.q)).collect();
        assert_eq!(R.mul(&a, &R.one()), a);
    }

    #[test]
    fn x_to_the_d_is_minus_one() {
        let R = Ring::new(TEST_PARAMS.q, 32);
        let mut x = vec![0u64; 32];
        x[1] = 1;
        let mut prod = R.one();
        for _ in 0..32 {
            prod = R.mul(&prod, &x);
        }
        let mut expected = vec![0u64; 32];
        expected[0] = R.q - 1;
        assert_eq!(prod, expected);
    }

    #[test]
    fn ntt_agrees_with_schoolbook_d128() {
        let R = Ring::new(PRODUCTION_PARAMS.q, 128);
        let mut r = lcg(3);
        for _ in 0..5 {
            let a: Poly = (0..128).map(|_| r(R.q)).collect();
            let b: Poly = (0..128).map(|_| r(R.q)).collect();
            assert_eq!(R.mul(&a, &b), R.mul_schoolbook(&a, &b));
        }
    }

    #[test]
    fn ntt_mat_vec_agrees_with_plain_mat_vec() {
        let R = Ring::new(PRODUCTION_PARAMS.q, 128);
        let mut r = lcg(5);
        let mat: PolyMat = (0..3)
            .map(|_| (0..4).map(|_| (0..128).map(|_| r(R.q)).collect()).collect())
            .collect();
        let v: PolyVec = (0..4).map(|_| (0..128).map(|_| r(R.q)).collect()).collect();
        let mat_ntt = R.mat_to_ntt(&mat).expect("production ring has CRT backend");
        assert_eq!(R.mat_vec_ntt(&mat_ntt, &v).unwrap(), R.mat_vec(&mat, &v));
    }

    #[test]
    fn centered_decompose_roundtrip() {
        let R = Ring::new(PRODUCTION_PARAMS.q, 128);
        let mut r = lcg(4);
        for k_bits in [4u32, 8, 12] {
            let a: Poly = (0..128).map(|_| r(R.q)).collect();
            let (hi, lo) = R.centered_decompose(&a, k_bits);
            let reconstructed = R.add(&R.scale(1i64 << k_bits, &hi), &lo);
            assert_eq!(reconstructed, a);
            assert!(R.inf_norm(&lo) <= 1u64 << (k_bits - 1));
        }
    }
}
